pkg/lib/indexer/indexer.go

```diff
@@ -283,3 +292,190 @@ func write(dockerfileText, outDockerfile string, logger *logrus.Entry) error {
 
 	return nil
 }
+
+// ExportFromIndexRequest defines the parameters to send to the ExportFromIndex API
+type ExportFromIndexRequest struct {
+	Index         string
+	Package       string
+	DownloadPath  string
+	ContainerTool string
+}
+
+// ExportFromIndex is an aggregate API used to specify operators from
+// an index image
+func (i ImageIndexer) ExportFromIndex(request ExportFromIndexRequest) error {
+	databaseFile := defaultDatabaseFile
+
+	// set a temp directory
+	workingDir, err := ioutil.TempDir("./", "index_tmp")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(workingDir)
+
+	// extract the index database to the file
+	databaseFile, err = i.WriteIndexDBFile(request.Index, workingDir, databaseFile)
+	if err != nil {
+		return err
+	}
+
+	bundles, err := getBundlesToExport(databaseFile, request.Package)
+	if err != nil {
+		return err
+	}
+	i.Logger.Infof("Preparing to pull bundles %+q", bundles)
+
+	// Creating downloadPath dir
+	if _, err := os.Stat(request.DownloadPath); err != nil {
+		if os.IsNotExist(err) {
+			err = os.Mkdir(request.DownloadPath, 0777)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	var errs []error
+
+	db, err := sql.Open("sqlite3", databaseFile)
+	if err != nil {
+		return err
+	}
+
+	dbQuerier := sqlite.NewSQLLiteQuerierFromDb(db)
+	if err != nil {
+		return err
+	}
+
+	for _, bundleImage := range bundles {
+		// try to name the folder
+		folderName, err := dbQuerier.GetBundleVersion(context.TODO(), bundleImage)
+		if err != nil {
+			return err
+		}
+		if folderName == "" {
+			// operator-registry does not care about the folder name
+			folderName = bundleImage
+		}
+		exporter := bundle.NewSQLExporterForBundle(bundleImage, filepath.Join(request.DownloadPath, folderName), request.ContainerTool)
+		if err := exporter.Export(); err != nil {
+			err = fmt.Errorf("error exporting bundle from image: %s", err)
+			errs = append(errs, err)
+		}
+	}
+	err = db.Close()
+	if err != nil {
+		errs = append(errs, err)
+		return utilerrors.NewAggregate(errs)
+	}
+
+	err = generatePackageYaml(databaseFile, request.Package, request.DownloadPath)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	return utilerrors.NewAggregate(errs)
+}
+
+func (i ImageIndexer) WriteIndexDBFile(index, workingDir, databaseFile string) (string, error) {
```

Comment:
It's a little confusing to have two modes of operation based on the value of `index`, one of which doesn't write a file, and the other doesn't use the value of `databaseFile`. The behavior of this function might be clearer if it were narrowed down to lines 381-398 and only accepted `index` and `workingDir`. Then the caller can decide what to do if it doesn't have a value for `index`.

Reply:
I agree this function is confusing, I just pulled the code out of the the main `ExportFromIndex()` function into this to make it more clear. I started to refactor it but would like to write a test for it first to ensure behavior, it's sort of tricky.